text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,315 @@
+
+### Explanation
+
+In order to use the tokenizer, all the user has to do is to instantiate a `Tokenizer`, call `tokenize(&str)` on it and iterate over the emitted tokens:
+
+```rust
+let stop_words = Set::from_iter(["the", "of"].iter()).unwrap();
+let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));
+
+let orig = "The quick (\"brown\") fox can't jump 32.3 feet, right? Brr, it's 29.3°F!";
+let analyzed = analyzer.analyze(orig);
+let mut analyzed = analyzed.tokens();
+assert_eq!("the", analyzed.next().unwrap().text());
+assert_eq!(" ", analyzed.next().unwrap().text());
+assert_eq!("quick", analyzed.next().unwrap().text());
+assert_eq!(" ", analyzed.next().unwrap().text());
+assert_eq!("(", analyzed.next().unwrap().text());
+assert_eq!("\"", analyzed.next().unwrap().text());
+assert_eq!("brown", analyzed.next().unwrap().text());
+```
+
+The call to the tokenize method allows the reuse of the same `Tokenizer` instance, and keep it's configuration state and allocations.
+
+Bellow are example of the integration of the new tokenizer in existing code:
+
+- Highlight in @kerollmops milli:
+```rust
+// new tokenizer
+fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
+   let stop_words = Set::default();
+    let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));
+
+    for (_key, value) in record.iter_mut() {
+        let old_value = mem::take(value);
+        let analyzed = analyzer.analyze(&old_value);
+        
+        for (original, token) in analyzed.reconstruct() {
+            if token.is_word() {
+                let to_highlight = words.contains(&token.text());
+                if to_highlight { value.push_str("<mark>") }
+                value.push_str(original);
+                if to_highlight { value.push_str("</mark>") }
+            } else {
+                value.push_str(original);
+            }
+        }
+    }
+}
+```
+```rust
+// original
+fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
+    for (_key, value) in record.iter_mut() {
+        let old_value = mem::take(value);
+        for (token_type, token) in simple_tokenizer(&old_value) {
+            if token_type == TokenType::Word {
+                let lowercase_token = token.to_lowercase();
+                let to_highlight = words.contains(&lowercase_token);
+                if to_highlight { value.push_str("<mark>") }
+                value.push_str(token);
+                if to_highlight { value.push_str("</mark>") }
+            } else {
+                value.push_str(token);
+            }
+        }
+    }
+}
+```
+
+As we can see, the changes that need to be made are very minimal: this is because efforts have been made to make it's API close to the previous one.
```

Comment:
```suggestion
As we can see, the changes that need to be made are very minimal: this is because efforts have been made to make its API close to the previous one.
```